### DISCUSSION TITLE
Improvement help

### DISCUSSION THREAD

**Discussion by MattBystrin** [premake/premake-core#1804] [Q&A] [answered]
Hello and thank you for a great software. It's a big pleasure using it. 
I'm want to reforge an old project https://github.com/tarruda/premake-export-compile-commands.  I knew what some changes been made, but still got an improvements. But I'm faced some problems:
1) How to get default values ( toolset, config and etc ) that is going to be using invoking premake. The first one on the list ? In example, in old module I saw such code `return p.tools[cfg.toolset or 'gcc']` that seems to me confusing. What if gcc is not system default?
2) I didn't find some doc for developer, which could be really nice. I know it takes plenty of time, unfortunally. But doing this module I can participate in doc creation.

**Comment by samsinsane** [comment, marked as answer]
> How to get default values ( toolset, config and etc ) that is going to be using invoking premake.

Default values are baked into the `wks`, `prj` and `cfg` objects during the generation process. In the case of the module that you've linked, the `export-compile-commands` doesn't specify a default toolset, so it will be `nil` unless the user explicitly sets it, runs on a Darwin-based OS (Premake sets this globally) or you default it at the point you require it (as is done in the module currently).

> What if gcc is not system default?

Then it won't work as expected. Users on systems with another compiler as the default should specify the toolset. I believe there's a Premake option for this already, `--cc=clang` for example. 

> I didn't find some doc for developer

There's the general docs [here](https://premake.github.io/docs/), which contains a whole section on extending Premake, [here](https://premake.github.io/docs/#extending-premake).

**Reply by MattBystrin** [reply]
Thanks for your reply !
